pkg/upstream/rewrite.go

```diff
@@ -50,6 +50,7 @@ func rewritePath(rewriteRegExp *regexp.Regexp, rewriteTarget string, writer page
 		}
 
 		req.RequestURI = reqURL.String()
+		req.URL.Path = reqURL.Path // set path for websocket connections
```

Comment:
Have you tested the impact of this change to "normal" network connections?

Reply:
Yes, I did test both websocket as well as non websocket connections using the following oauth2-proxy config:

```yaml
providers:
- clientID: account
  clientSecret: dummy
  id: oidc=account
  oidcConfig:
    emailClaim: email
    groupsClaim: groups
    insecureSkipNonce: true
    issuerURL: http://localhost:8080/realms/master
    userIDClaim: email
    audienceClaims: ['aud']
  provider: oidc
server:
  BindAddress: 127.0.0.1:8081
  SecureBindAddress: ""
  TLS: null
upstreamConfig:
  upstreams:
  - flushInterval: 1s
    id: prod
    passHostHeader: true
    path: ^/prod/(.*)$
    rewriteTarget: /$1
    proxyWebSockets: true
    uri: http://127.0.10.0:8080/
    timeout: 600s
  - flushInterval: 1s
    id: staging
    passHostHeader: true
    path: ^/staging/(.*)$
    rewriteTarget: /$1
    proxyWebSockets: true
    uri: http://127.0.11.0:8080/
    timeout: 600s
```

```bash
OAUTH2_PROXY_COOKIE_SECRET="<REDACTED>" ./oauth2-proxy --alpha-config ~/config.yml --email-domain="*" --skip-jwt-bearer-tokens=true --extra-jwt-issuers=http://localhost:8080/realms/master=account
```

And the following python code:

```python
import requests
from oauthlib.oauth2 import BackendApplicationClient
from requests_oauthlib import OAuth2Session

# OAuth provider configuration
client_id = 'test-client'
client_secret = '<REDACTED>'
token_url = 'http://localhost:8080/realms/master/protocol/openid-connect/token' 

# Create an OAuth2Session with the BackendApplicationClient
client = BackendApplicationClient(client_id=client_id)
oauth = OAuth2Session(client=client)

# Fetch the access token using client credentials grant
token = oauth.fetch_token(
    token_url=token_url,
    client_id=client_id,
    client_secret=client_secret,
    scope='openid profile' 
)

headers = {"Authorization": f"Bearer {token['access_token']}"}
headers_ws = {"Authorization": f"Bearer {token['access_token']}", "Upgrade": "websocket", "Connection": "Upgrade", "Sec-WebSocket-Key": "SGVsbG8sIHdvcmxkIQ==", "Sec-WebSocket-Version": "13"}


print("##### PROD ######")
response = requests.get('http://localhost:8081/prod/test123?foo=bar', headers=headers)
print(response.content.decode())

print("##### STAGING ######")
response_stg = requests.get('http://localhost:8081/staging/test123?foo=bar', headers=headers)
print(response_stg.content.decode())

print("##### PROD WS ######")
response_ws = requests.get('http://localhost:8081/prod/test123?foo=bar', headers=headers_ws)
print(response_ws.content.decode())

print("##### STAGING WS ######")
response_ws = requests.get('http://localhost:8081/staging/test123?foo=bar', headers=headers_ws)
print(response_ws.content.decode())
```

I had the `jmalloc/echo-server` image listening at `127.0.10.0:8080` and `127.0.11.0:8080`. These are the requests as seen by the echo-server:

Before my modification:

```
10.0.2.100:41568 | GET /test123?foo=bar
10.0.2.100:53328 | GET /prod/test123?foo=bar
10.0.2.100:53328 | upgraded to websocket
10.0.2.100:53328 | websocket: close 1006 (abnormal closure): unexpected EOF
```

```
10.0.2.100:41576 | GET /test123?foo=bar
10.0.2.100:53340 | GET /staging/test123?foo=bar
10.0.2.100:53340 | upgraded to websocket
10.0.2.100:53340 | websocket: close 1006 (abnormal closure): unexpected EOF
```

After my modification:

```
10.0.2.100:56578 | GET /test123?foo=bar
10.0.2.100:56596 | GET /test123?foo=bar
10.0.2.100:56596 | upgraded to websocket
10.0.2.100:56596 | websocket: close 1006 (abnormal closure): unexpected EOF
```

```
10.0.2.100:56590 | GET /test123?foo=bar
10.0.2.100:56602 | GET /test123?foo=bar
10.0.2.100:56602 | upgraded to websocket
10.0.2.100:56602 | websocket: close 1006 (abnormal closure): unexpected EOF
```

Let me know if there are more tests I could run.

Reply:
Thanks for the detailed answer! Much appreciated.

Reply:
Why does the staging example change with this modification? I'm not following why that changed, nor why the gap-signature change. If the gap-signature changes that suggests to me the path that is hitting the backend has changed, which might constitute a breaking change

Reply:
The staging example changes because I have the following upstream config:

```
  - flushInterval: 1s
    id: staging
    passHostHeader: true
    path: ^/staging/(.*)$
    rewriteTarget: /$1
    proxyWebSockets: true
    uri: http://127.0.11.0:8080/
    timeout: 600s
````

The gap signature changes because when the signature is calculated `req.URL.Path` is taken into consideration: 

https://github.com/mbland/hmacauth/blob/44256dfd4bfa7594cfa73587a464ca890e85971a/hmacauth.go#L130

This is not the case for `req.RequestURI` so no matter how you change `req.RequestURI` the gap signature won't change. There is no way I can modify `req.URL.Path` without the gap-signature changing.